### DISCUSSION TITLE
How to  use function code 0x06 to read from register?

### DISCUSSION THREAD

**Discussion by ruteclrp** [pymodbus-dev/pymodbus#2066] [Help] [answered]
I have a heatpump where the implementation of Modbus appear somewhat unconvential. Case is that for each register type (function code) the starting address is identical.
For example - for function code 0x04:
Request:
Function code                                1 byte              0x04
Starting address                             2 bytes            0x0001 – 0x0014 (20)
Quantity of input registers             2 bytes            0x0001
Response:
Function code                                1 byte              0x04
Byte count                                      1 byte              0x02
Sensor value registers                    2 bytes             0xHiLo (sensor values 1-20)

For function code 0x06:
Request:
Function code                                1 byte              0x06
Register address                             2 bytes            0x0001 – 0x00D2 (*1)
Register value                                 2 bytes            0x0000 – 0x00FF (*2)
Response:
Function code                                 1 byte             0x06
Register address                              2 bytes           0x0001 – 0x00D2
Register value                                  2 – 4 bytes     Depending on msg. See list of variables.
(*1) Add 0x0100 to Register address for writing request (only for address 0x0001 – 0x0091 )
(*2) depending on each address min/max area. Value is Do-Not-Care on reading only.

I can write using the function code 0x06 by adding 0x0100 (256) to each register address.

I can however not read the content of the registers under function code 0x06, as that requires using the write function to the lower address space and ignoring the write payload and rather take the response values and treat them as if I was using function code 0x03 or 0x04.

How could I possibly read those addresses and process the response using the write function 0x06?

appreciate anyone with deeper insight in pymodbus.

**Comment by janiversen** [comment]
There are a lot of devices that do not have shared address space, so you treat the 4 different blocks as individual address spaces. Please see the simulator datastore simulator documentation where this is explained.

I am not sure I understand your problem, especially not the mix with write.

Seems your device uses a different address to write than to read, that is somewhat very old fashioned, but works....you cannot read the "write address" you read the lower address and write the upper address...when writing the result should be in the lower address (depending on the type of register, e.g. set_temperature, write sets the temperature but read reads the actual temperature.

This is not a pymodbus problem, but how to handle your device, and that is something you need to read in the manual of the device.

**Comment by janiversen** [comment]
And just for the record 0x06 can pr definition not READ...it's a write function.

**Reply by ruteclrp** [reply]
That was what I thought too. However, when you se the specification for using function code 0x06 above you will notice that "writing" to the lower addresses, the register value parsed with the write function is "do not care", and the response from the write function carries the register value as stored in that register.

So in essence, function code 0x06 is a read function when writing to the lower addresses.

Now, what I need is obviously a way to use the response from such write function to be treated as if I used function code 0x03 or 0x04.

And, quoting from the heatpump's modbus specification: 
Supported function codes:
0x01 – Read coils
0x02 – read discrete inputs
0x03 – read holding registers
0x04 – read input registers
0x06 – write single register **(tweaked to also just read)**

And, as the first register under function code has the address 0x01, this is the address for reading the value which is stored by writing to the address 0x0101. Both addresses can only be reached using function code 0x06.

So, in essence what I need is a way to use the response arising out of writing some arbitrary value (which is ignored) to 0x01 using function code 0x06, and be able to take the response in the format of: 0x06 0x01 0xXX and parse that last 2 byte as my read value.

I hope this clarifies

ps. to make matter worse the need is in Home Assistant where there seems even less options to deviate from the standard methods.

**Reply by janiversen** [reply]
it requires you tweak pymodbus for the 0x06 response, it's possible but not a very easy task.

And yes HA is clearly more restrictive (I am the main maintainer), there you would need to tweak quite a lot in the modbus integration.

Possible the best method would be to tweak the read call in pymodbus, so that it issues a 0x06 when needed..that way you would avoid having to tweak the HA modbus integration.

0x06 0x01 0xXX, is not even a correct response, 0x01 signals 1 register which is 2bytes not 1. The expected response is 0x06 0xAA 0xAA 0xBB 0xBB where AAAA is address and BBBB is the register value. So it looks a lot like you have a different problem.

**Reply by janiversen** [reply]
Just checked pymodbus delivers the register value in the return, when using 0x06, so no need for tweaking seen from pymodbus pow.

I think the manual tries to say if you issue 0x06, there are no write but the values is returned as normal..,,so no breaking of the modbus protocol.

**Reply by ruteclrp** [reply]
Thanks a lot. I know the response I wrote was a bit sloppy - however it was more to indicate what I was talking about. Naturally, the register is 2 bytes, and the value also 2 byte.

Anyway, just to understand your response. I read it as you are proposing adding a new "read" method which build on e.g. 0x04 calls but uses 0x06 in the execution of the "read" call.
Ideally I would need an additional input_type in HA, such as "input_6" which would work just as input type "input" but call using 0x06 rather than 0x04. I wonder how difficult that would be. In any case, in HA I would need some way to tell the integration that the "read" call should use function 0x06.

To be perfectly honest here, I am in deep water here :-)

**Reply by ruteclrp** [reply]
So, on your second reply above, it is really the parsing of the returned register value within HA which is the issue, and possibly, adding a new "input_6" method to the input type options could make it possible for the standard "input" method to just used 0x06 rather than 0x04 in the call.

**Reply by janiversen** [reply, marked as answer]
well at least it is not a pymodbus issue….and in HA it is a lot more than just adding a new input_type, it’s tweaking a fair amount of code.

It is probably easier to tweak the read request/response in pymodbus and keep HA unchanged. Or even easier, make a new library that sits between pymodbus and HA that does the tweaking you need…anyhow it is up to you.
